Normalize arithmetic equalities.

blishko opened the pull request:
Depends on #292.
Resolves #267.

Currently, arithmetic equalities are not normalized, they are created as given, in the same way as equalities for uninterpreted sorts.
This PR adds normalization process to arithmetic equalities on creation, which works almost the same as the normalization for arithmetic inequalities.

aehyvari commented on src/logics/LIALogic.cc:
```diff
@@ -201,8 +212,19 @@ PTRef LIALogic::sumToNormalizedInequality(PTRef sum) {
+
+
+PTRef LIALogic::sumToNormalizedInequality(PTRef sum) {
+    auto[lhsVal, rhs] = sumToNormalizedPair(sum);
```
missing a space

blishko replied:
Fixed.

aehyvari commented on src/logics/LIALogic.cc:
```diff
@@ -201,8 +212,19 @@ PTRef LIALogic::sumToNormalizedInequality(PTRef sum) {
+}
+
+PTRef LIALogic::sumToNormalizedEquality(PTRef sum) {
+    auto[lhsVal, rhs] = sumToNormalizedPair(sum);
```
Missing a space

blishko replied:
Fixed.

aehyvari commented on src/logics/LIALogic.cc:
```diff
@@ -201,8 +212,19 @@ PTRef LIALogic::sumToNormalizedInequality(PTRef sum) {
     // in LIA we can strengthen the inequality to
     // ceiling(-constantValue) <= normalizedSum
```
On line 2010 above we assume that the sum is normalised, which should indeed be the case with the current code.  I wonder would it make sense to write an assert for that?

blishko replied:
Yes, the comment is obsolete, I'll remove it.

> On line 210 above we assume that the sum is normalised, which should indeed be the case with the current code. I wonder would it make sense to write an assert for that?

Yes, but how should it look like? Just `assert(normalizedSum == mkNumPlus(varFactors)`? Or do you have something else in mind?

aehyvari reviewed (commented):
This looks great!

I tried out a refactoring where I put to a function `std::pair<FastRational, vec<PTRef>>LALogic::getConstantAndFactors(PTRef)` the common part of `LIALogic::sumToNormalizedPair` and `LRALogic::sumToNormalizedPair`.  If you want I can push it here.

aehyvari commented on src/logics/LRALogic.cc:
```diff
@@ -161,4 +163,58 @@ PTRef LRALogic::insertTerm(SymRef sym, vec<PTRef> &&terms) {
+    splitTermToVarAndConst(leadingFactor, var, coeff);
+    opensmt::Number normalizationCoeff = abs(getNumConst(coeff));
+    // varFactors come from a normalized sum, no need to call normalization code again
+    PTRef normalizedSum = varFactors.size() == 1 ? varFactors[0] : mkFun(get_sym_Num_PLUS(), std::move(varFactors));
```
(see the related comment above for `LIALogic`)

blishko replied:
Yeah. the idea is to avoid normalization process when we know it's not needed.

blishko commented on src/logics/LRALogic.cc:
```diff
@@ -161,4 +163,58 @@ PTRef LRALogic::insertTerm(SymRef sym, vec<PTRef> &&terms) {
+ * it was placed on the other side of an equality.
+ *
+ * @param sum
+ * @return Constant part of the normalized sum as LHS and non-constant part of the normalized sum as RHS
```
You are right. I think I originally wanted to have a pair of PTRefs and then realized than in LIA I need the Rational value rather than PTRef. It seems to be fine to unify it in the way you suggest.

blishko commented on src/logics/LIALogic.cc:
```diff
@@ -116,10 +116,21 @@ LIALogic::LIALogic() :
+    assert(isNumPlus(sum));
     vec<PTRef> varFactors;
     PTRef constant = PTRef_Undef;
     Pterm const & s = getPterm(sum);
```
Fixed.

blishko commented on src/logics/LRALogic.cc:
```diff
@@ -161,4 +163,58 @@ PTRef LRALogic::insertTerm(SymRef sym, vec<PTRef> &&terms) {
+    vec<PTRef> varFactors;
+    PTRef constant = PTRef_Undef;
+    Pterm const & s = getPterm(sum);
+    for (int i = 0; i < s.size(); i++) {
```
Fixed.

blishko commented:
I think I addressed all the comments. There is the question of adding an assertion for normalized sum.

> I tried out a refactoring where I put to a function std::pair<FastRational, vec<PTRef>>LALogic::getConstantAndFactors(PTRef) the common part of LIALogic::sumToNormalizedPair and LRALogic::sumToNormalizedPair. If you want I can push it here.

Yes, there seems to be some common code. You can add it here. Or you can create a separate branch for that and create a PR to my PR :)

blishko commented on src/logics/LIALogic.cc:
```diff
@@ -188,9 +175,10 @@ opensmt::pair<FastRational, PTRef> LIALogic::sumToNormalizedPair(PTRef sum) {
-    for (std::size_t i = 1; i < coeffs.size() && coeffs_gcd != 1; ++i) {
-        coeffs_gcd = gcd(coeffs_gcd, abs(coeffs[i]));
+    for (auto const & coeff : coeffs) {
+        coeffs_gcd = gcd(coeffs_gcd, abs(coeff));
```
This now adds one unnecessary `gcd` operation of the first coefficient with itself.
I think I would rather stick to the original version.

aehyvari replied:
Right, it slipped past me, I didn't want to commit it.  I'll take it off.

blishko replied:
OK. Otherwise the refactoring is nice!

aehyvari commented on src/logics/LRALogic.cc:
```diff
@@ -173,25 +173,11 @@ PTRef LRALogic::insertTerm(SymRef sym, vec<PTRef> &&terms) {
-    opensmt::Number constantVal = getNumConst(constant);
-    assert(varFactors.size() > 0);
-    termSort(varFactors);
+opensmt::pair<FastRational, PTRef> LRALogic::sumToNormalizedPair(PTRef sum) {
```
I'm going to change this into 

`opensmt::pair<opensmt::Number, PTRef> LRALogic::sumToNormalizedPair(PTRef sum) {`

blishko replied:
OK!

blishko commented:
Do you want to add assertions to the normalized sums? Or can we merge the current state?

aehyvari commented:
> Do you want to add assertions to the normalized sums? Or can we merge the current state?

I think we can merge the current state.
